File: mpp/lab3/src/main/java/repository/AbstractRepository.java
package repository;

import domain.HasID;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public abstract class AbstractRepository<E extends HasID<ID>, ID> implements Repository<E, ID> {
    protected HashMap<ID, E> map = new HashMap<ID, E>();
    private Validator<E> val;

    AbstractRepository(Validator<E> _val) {
        this.val = _val;
    }

    @Override
    public long size() {
        return this.map.size();
    }

    @Override
    public E save(E entity) throws ValidationException {
        this.val.validate(entity);

        if (this.map.containsKey(entity.getId()))
            return this.map.get(entity.getId());

        return this.map.put(entity.getId(), entity);
    }

    @Override
    public E delete(ID id) {
        return this.map.remove(id);
    }

    @Override
    public E findOne(ID id) {
        return this.map.get(id);
    }

    @Override
    public Map<ID, E> getAll() {
        return this.map;
    }

    @Override
    public Iterable<E> findAll() {
        return this.map.values();
    }

    @Override
    public E update(E entity) throws ValidationException {
        this.val.validate(entity);

        if (this.map.containsKey(entity.getId())) {
            this.map.put(entity.getId(), entity);
            return null;
        }

        return entity;
    }
}

File: mpp/proj1_real/WindowsFormsApp1/WindowsFormsApp1/Repository/AbstractRepository.cs
using System.Collections.Generic;
using WindowsFormsApp1.Model;

namespace WindowsFormsApp1.Repository
{
    public abstract class AbstractRepository<ID, T> : IRepository<ID, T> where T : IHasId<ID>
    {
        IDictionary<ID, T> items;

        public AbstractRepository(IValidator<T> validator)
        {
            items = new Dictionary<ID, T>();

        }

        public AbstractRepository()
        {
            items = new Dictionary<ID, T>();
        }


        public virtual void delete(ID id)
        {
            items.Remove(id);

        }

        public IEnumerable<T> findAll()
        {
            return items.Values;
        }

        public T findOne(ID id)
        {
            if (items.ContainsKey(id))
                return items[id];
            else
                return default(T);
        }

        public virtual void save(T entity)
        {
            if (!items.ContainsKey(entity.Id))
                items.Add(entity.Id, entity);
            else throw new RepositoryException("Duplicate entity " + entity);
        }
    }
}